### PR TITLE
JDBC executeUpdate support

### PR DESCRIPTION
**Pull request by losipiuk**
Replacement for https://github.com/TeradataCenterForHadoop/presto-td/pull/2

### CODE REVIEW

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 1]
```diff
@@ -170,10 +166,62 @@ public void setCursorName(String name)
...
+
+        StatementClient client = null;
+        ResultSet resultSet = null;
+        try {
+            client = connection().startQuery(sql);
+            resultSet = new PrestoResultSet(client);
+            QueryResults queryResults = client.current();
+
+            // query statement
+            if (queryResults.getUpdateType() == null) {
```
Comment (kokosing) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: check queryResults.getStats().getState() in case that query failed so you could throw some SQLException instead of returning empty result set

**Comment by kokosing** [comment]
@electrum Comment applied.

**Comment by electrum** [comment]
Can you squash these together? I'm not sure who the author should be.

**Comment by kokosing** [comment]
I squashed them, I selected @sogorkis  to be an author as IMO did the most of this patch.
I also rebased and fixed some tests.

**Comment by kokosing** [comment]
It appeared that checking the error in prepared statement revealed some intermittent error. I do not know  why it happens from time to time only and not always. I pushed a fix in this PR. See log of mentioned error occurence: https://s3.amazonaws.com/archive.travis-ci.org/jobs/84457475/log.txt

**Comment by kokosing** [comment]
Travis build failed, although it is passing on my sandbox.

**Comment by kokosing** [comment]
rebased

**Comment by losipiuk** [comment]
Rebased on top of master. 

@electrum can you take a look?

I kinda do not like how this patch validates if proper method `executeQuery`/`executeUpdate` was used for `SELECT`/`INSERT` query. It is currently verified after query is sent to Presto for execution. 

Probably better solution would be to prevalidate it in driver itself.
Or pass the request type (QUERY/UPDATE) to presto in initial `/v1/statement` request and do validation on Presto side.

I personally lean toward second solution so we keep driver logic as simple as possible and keep query parsing stuff only in Presto server codebase.

What do you think?

cc: @kokosing, @ilfrin, @sopel39

**Comment by kokosing** [comment]
@losipiuk second option sounds good to me

**Comment by kokosing** [comment]
Rebased and unit tests fixes.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 2]
```diff
@@ -45,21 +51,11 @@
...
             throws SQLException
     {
-        try {
-            ResultSet result = new PrestoResultSet(connection().startQuery(sql));
-            currentResult.set(result);
-            return result;
-        }
-        catch (RuntimeException e) {
-            throw new SQLException("Error executing query", e);
+        boolean isQueryStatement = execute(sql);
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: No need for an extra variable here

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 3]
```diff
@@ -45,21 +51,11 @@
...
+        if (!isQueryStatement) {
+            throw new SQLException("Not an query statement: " + sql);
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: This would be better phrased as `"SQL statement is not a query"`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 4]
```diff
@@ -170,10 +166,66 @@ public void setCursorName(String name)
     public boolean execute(String sql)
             throws SQLException
     {
+        clearCurrentResults();
+
         checkOpen();
-        // Only support returning a single result set
-        currentResult.set(executeQuery(sql));
-        return true;
+        requireNonNull(sql, "null sql statement");
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: No need to check for null here -- already checked in `StatementClient`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 5]
```diff
@@ -170,10 +166,66 @@ public void setCursorName(String name)
...
+
+        StatementClient client = null;
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: Use try-with-resources

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 6]
```diff
@@ -170,10 +166,66 @@ public void setCursorName(String name)
...
+        ResultSet resultSet = null;
+        try {
+            client = connection().startQuery(sql);
+            resultSet = new PrestoResultSet(client);
+            QueryResults queryResults = client.current();
+
+            if (queryResults.getError() != null) {
+                throw new SQLException(queryResults.getError().getMessage());
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: This should use `resultsException()` in `PrestoResultSet`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 7]
```diff
@@ -170,10 +166,66 @@ public void setCursorName(String name)
...
+        requireNonNull(sql, "null sql statement");
+
+        StatementClient client = null;
+        ResultSet resultSet = null;
+        try {
+            client = connection().startQuery(sql);
+            resultSet = new PrestoResultSet(client);
+            QueryResults queryResults = client.current();
+
+            if (queryResults.getError() != null) {
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: Use `client.isFailed()`, then call `client.finalResults()`

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 8]
```diff
@@ -170,10 +166,66 @@ public void setCursorName(String name)
...
+                return true;
+            }
+
+            // update statement
+            while (resultSet.next()) {
+                // we just care about update count
+            }
+            queryResults = client.finalResults();
+
+            Long queryUpdateCount = queryResults.getUpdateCount();
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]:
```java
Long updateCount = client.finalResults().getUpdateCount();
currentUpdateCount.set((updateCount != null) ? updateCount : 0);

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 9]
```diff
@@ -189,8 +241,15 @@ public int getUpdateCount()
             throws SQLException
     {
         checkOpen();
-        // Updates are not allowed yet so return -1
-        return -1;
+        return Ints.checkedCast(getLargeUpdateCount());
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: Throwing `IllegalArgumentException` is definitely wrong here (this method doesn't take any arguments).

I think the two choices are `Ints.saturatedCast` which will return `Integer.MAX_VALUE` or throwing a `SQLException`. @dain and I agree that the former seems like a better choice.

Do you know what other drivers return here? For example, the Teradata driver? I remember that Netezza driver from 2010 threw an exception, which made it impossible to use JDBC for batch ETL (and this was long before `getLargeUpdateCount()` was added in Java 1.8).

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 10]
```diff
@@ -313,45 +372,84 @@ public ResultSet getGeneratedKeys()
     }
 
     @Override
+    public int executeUpdate(String sql)
+            throws SQLException
+    {
+        return Ints.saturatedCast(executeLargeUpdate(sql));
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: Same comment as above. The behavior should definitely be consistent across these related methods.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 11]
```diff
@@ -313,45 +372,84 @@ public ResultSet getGeneratedKeys()
...
-        throw new SQLFeatureNotSupportedException("executeUpdate");
+        return executeUpdate(sql);
+    }
+
+    @Override
+    public long executeLargeUpdate(String sql)
+            throws SQLException
+    {
+        boolean isQueryStatement = execute(sql);
+        if (isQueryStatement) {
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: No need for a variable

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java [hunk 12]
```diff
@@ -313,45 +372,84 @@ public ResultSet getGeneratedKeys()
...
+            throw new SQLException("Not an update statement: " + sql);
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: `"SQL statement is a query (not an update)"`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 13]
```diff
@@ -69,6 +71,22 @@ public void setup()
...
+        server.createCatalog("blackhole", "blackhole");
+
+        setupTestTables();
+    }
+
+    private void setupTestTables()
+            throws SQLException
+    {
+        try (
+                Connection connection = createConnection("blackhole", "blackhole");
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Format like this:
```java
 try (Connection connection = createConnection("blackhole", "blackhole");
        Statement statement = connection.createStatement()) {

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 14]
```diff
@@ -69,6 +71,22 @@ public void setup()
...
+                Statement statement = connection.createStatement()
+        ) {
+            int updated = statement.executeUpdate("CREATE TABLE test_table (key BIGINT)");
+            assertGreaterThanOrEqual(updated, 0);
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: This should be exactly zero:
```
assertEquals(updated, 0);

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 15]
```diff
@@ -242,7 +260,7 @@ public void testGetCatalogs()
     {
         try (Connection connection = createConnection()) {
             try (ResultSet rs = connection.getMetaData().getCatalogs()) {
-                assertEquals(readRows(rs), list(list("system"), list(TEST_CATALOG)));
+                assertEquals(readRows(rs), list(list("blackhole"), list("system"), list(TEST_CATALOG)));
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Please split this out into a separate commit: "Add blackhole connector to JDBC tests"

This will make it easier to see which test changes are directly related to `executeUpdate()` changes

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 16]
```diff
@@ -816,14 +846,17 @@ public void testGetSuperTypes()
...
 
     @Test
-    public void testExecute()
+    public void testExecuteWithQuery()
             throws Exception
     {
         try (Connection connection = createConnection()) {
             try (Statement statement = connection.createStatement()) {
                 assertTrue(statement.execute("SELECT 123 x, 'foo' y, CAST(NULL AS bigint) z"));
                 ResultSet rs = statement.getResultSet();
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Let's do it like this:
```java
ResultSet rs = statement.getResultSet();
assertEquals(statement.getUpdateCount(), -1);
assertEquals(statement.getLargeUpdateCount(), -1);
assertTrue(rs.next());

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 17]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+    public void testExecuteWithInsert()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole")) {
+            try (Statement statement = connection.createStatement()) {
+                assertFalse(statement.execute("INSERT INTO test_table VALUES (1)"));
+
+                assertNull(statement.getResultSet());
+                assertEquals(statement.getUpdateCount(), 1);
+                assertEquals(statement.getLargeUpdateCount(), 1L);
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: All of these `getLargeUpdateCount()` checks can use `0` instead of `0L` since they are compared against a primitive long. The `0L` is only needed if the other side is an object (in which case the `0` would incorrectly box to `Integer`).

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 18]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+        }
+    }
+
+    @Test
+    public void testExecuteUpdateWithDDLStatement()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole")) {
+            try (Statement statement = connection.createStatement()) {
+                assertEquals(statement.executeUpdate("INSERT INTO test_table VALUES (1)"), 1);
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Do multiple values here:
```java
assertEquals(statement.executeUpdate("INSERT INTO test_table VALUES (1), (2), (3)"), 3);

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 19]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+
+                assertNull(statement.getResultSet());
+                assertEquals(statement.getUpdateCount(), 1);
+                assertEquals(statement.getLargeUpdateCount(), 1L);
+            }
+        }
+    }
+
+    @Test
+    public void testExecuteUpdateWithDDLStatement()
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Our code style is to use camel case for words (including acronyms), even if they are normally capitalized differently. So this should be `testExecuteUpdateWithDDLStatement()` (even though it looks worse)

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 20]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+                assertEquals(statement.executeUpdate("INSERT INTO test_table VALUES (1)"), 1);
+                assertNull(statement.getResultSet());
+                assertEquals(statement.getUpdateCount(), 1);
+                assertEquals(statement.getLargeUpdateCount(), 1L);
+            }
+        }
+    }
+
+    @Test
+    public void testExecuteUpdateWithDMLStatement()
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: `testExecuteUpdateWithDmlStatement()`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 21]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+                }
+                catch (SQLException e) {
+                    assertEquals(e.getMessage(), "Not an update statement: SELECT 123 x, 'foo' y, CAST(NULL AS bigint) z");
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testExecuteQueryWithDDLStatement()
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: `testExecuteQueryWithDdlStatement()`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 19 above)
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: These method names are reversed: `INSERT` is DML, `CREATE` is DDL

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 22]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+        }
+    }
+
+    @Test
+    public void testExecuteUpdateWithDMLStatement()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole")) {
+            try (Statement statement = connection.createStatement()) {
+                assertEquals(statement.executeUpdate("CREATE TABLE test_table_2 (key BIGINT)"), 0);
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Name this table `test_execute_ddl`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 23]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+                }
+                catch (SQLException e) {
+                    assertEquals(e.getMessage(), "Not an query statement: INSERT INTO test_table VALUES (1)");
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testExecuteUpdateCountCurrentResultCleared()
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Let's name this `testStatementReuse`

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 24]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole")) {
+            try (Statement statement = connection.createStatement()) {
+                // update statement
+                assertFalse(statement.execute("INSERT INTO test_table VALUES (1)"));
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Use two values here

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java [hunk 25]
```diff
@@ -848,6 +881,111 @@ public void testExecute()
...
+
+                // query statement
+                assertTrue(statement.execute("SELECT 123 x, 'foo' y, CAST(NULL AS bigint) z"));
+                assertNotNull(statement.getResultSet());
+                assertEquals(statement.getUpdateCount(), -1);
+                assertEquals(statement.getLargeUpdateCount(), -1L);
+                statement.getResultSet().close();
+
+                // update statement
+                assertFalse(statement.execute("INSERT INTO test_table VALUES (1)"));
```
Comment (electrum) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: Use three values (so it is different than the above)

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 2 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 3 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 4 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 5 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done for StatementClient. Cant use for resultSet as it is closed only if currentResult was not set.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 7 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 6 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 8 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 9 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: I TD jdbc driver and it does saturated cast. I will the code that way.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 10 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 11 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 12 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 13 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 14 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 16 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 17 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 19 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: You mean that method should be called `testExecuteUpdateWithDdlStatement`, right? The name you suggested is the same as one which is in this PR :)

Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 18 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 20 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 22 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 21 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 23 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 24 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java (hunk 25 above)
Comment (losipiuk) [inline presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java]: done

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java (hunk 6 above)
Comment (losipiuk) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoStatement.java]: can't do - close is called conditionally.

**Comment by losipiuk** [comment]
I rebased onto master and applied comments.
Please take a look @electrum.

File: presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoResultSet.java [hunk 26]
```diff
@@ -1795,7 +1795,7 @@ private ResultsPageIterator(StatementClient client)
         }
     }
 
-    private static SQLException resultsException(QueryResults results)
+    public static SQLException resultsException(QueryResults results)
```
Comment (electrum) [inline presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoResultSet.java]: This should be package protected, not public

**Comment by electrum** [comment]
Merged, thanks!
